oauthenticator/github.py

```diff
@@ -113,6 +113,30 @@ def _github_client_secret_changed(self, name, old, new):
         config=True, help="Automatically allow members of selected organizations"
     )
 
+    fetch_teams = Bool(
+        False,
+        help="""
+        Fetch list of teams user is in and put it in auth_state.
+
+        'teams' will be a key in auth_state that has the same structure listed
+        in https://docs.github.com/en/rest/reference/teams#list-teams-for-the-authenticated-user.
+
+        Automatically adds read:org scope to requested scopes if it is not present.
+
+        Each user is currently limited to a maximum of 100 teams.
+        """,
+        config=True,
+    )
+
+    @validate('scope')
+    def _validate_scope(self, proposal):
+        """
+        Ensure read:org is requested if fetch_teams is true
+        """
+        if 'read:org' not in proposal.value:
+            return ['read:org'] + proposal.value
+        return proposal.value
```

Comment:
Automatically adding a scope is a change in behaviour, previously we've relied on administrators setting the scope themselves (e.g. to allow checking of private org memberships). On the other hand automatically setting it definitely makes this easier to configure and is a lot nicer. I'm undecided on what's best, though if we keep this as automatically set I think it'd be good to add something to the docs.

Reply:
Right now this logic is adding `read:org` scope independent on `fetch_teams` being configured or not which wouldn't make sense at least - right?

I've grown vary of convenience logic for admins that kind of need to understand what goes on anyhow, and how complicated it can become to document that, thinking that it overall leads to more complexity than benefits for admins using this software and us who maintains it.

I'm trying to make my vague hesitance towards automating this more explicit by considering a few situations a bit more explicitly.

1. The admin configuring this is aware of `read:org` will be appended because the admin has read the docs and understands this is implied by enabling `fetch_teams`.
   Then this is quite reasonable in my mind. But even then, after some time, even the admin who configured this initially and read the docs once could be confused about seeing this be requested by users after looking at docs observing it isn't explicitly declared in `scopes` or similar.
2. The admin configuring use of `fetch_teams` wasn't aware that `read:org` will be appended and got a functional installation directly without adjusting `scopes`.
   This admin may end up confused why users were asked for something more than explicitly declared in `scopes`. We may need to document this behavior also in `scopes` along with `fetch_teams` due to that.

Hmmm at this point I'm a more favorable towards documenting `fetch_teams` to be a feature that should be used with `read:org` scope, which should be manually configured alongside `fetch_teams`.

I'm even more favorable towards this now that I realize that this feature would also rely on `enable_auth_state` to be configured - right? See the JupyterHub documentation for authenticators: https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html#authentication-state. Z2JH provides a auth key by default, but it does not enable saving auth state by default.

Reply:
Good points! `read:org` potentially means the app has to be granted by the organisation owner in addition to the user. There's a couple of old issues somewhere in this repo where we tried to figure out exactly what GitHub organisation/user related scopes are required in different situations and it was very complicated.... and depended on the GitHub organisation and user's settings. Automatically adding one scope here to handle fetching teams solves one problem, but it's inconsistent with how the rest of this authenticator is configured.

cilogin handling things differently is a good point too, but I think it reflects how separately developed authenticators have been bundled together. It would definitely be nice to refactor them all to be consistent, or alternatively to decouple them into separate repos (we've already said no new authenticators will be accepted in this repo), but for now I think we'll have to accept they all use slightly different patterns.

Reply:
I feel fairly strongly that we shouldn't allow config that lets admins shoot themselves in the foot, and part of that is allowing config that has 0 chance of working under any circumstances. If enabling this feature without read:org works even in some capacity, then I think it can be optional. However, I don't think it is possible for this feature to work without read:org scope (I could be wrong here tho), so auto adding is the way to go as it prevents inconsistent config that doesn't actually work....

I understand that deciding which scopes are needed for what might be tricky. But if we enforce a simple rule - if a particular feature always needs a scope, we make sure that scope is requested when that feature is turned on.

However, I recognize that is inconsistent with how the authenticator works now. I also definitely don't want to get bogged down in this as part of this PR as it is quite ancillary. So I'll just remove this for now and document that that scope is needed.

Reply:
I had forgotten to only add it if `fetch_teams` is set. It's now done in 39b51b22e0e0cd7b6d6938415ed1a935728e2f9d. 

If I remove the code that automatically sets the correct scopes, and set `fetch_teams` to True, logging in produces this nasty error in the logs:

```
[E 2022-04-08 11:46:49.848 JupyterHub oauth2:380] Error fetching user teams 404 GET https://api.github.com/user/teams: {
     "documentation_url": "https://docs.github.com/rest/reference/teams#list-teams-for-the-authenticated-user",
     "message": "Not Found"
    }
[E 2022-04-08 11:46:49.850 JupyterHub web:1789] Uncaught exception GET /hub/oauth_callback?code=c1e50d7dbcbf5171fcf1&state=eyJzdGF0ZV9pZCI6ICJkOTNlNDdiYjQzZWU0N2I2YmYwZmEwOGU5OTI4MWZmMCIsICJuZXh0X3VybCI6ICIifQ%3D%3D (::ffff:127.0.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8000', method='GET', uri='/hub/oauth_callback?code=c1e50d7dbcbf5171fcf1&state=eyJzdGF0ZV9pZCI6ICJkOTNlNDdiYjQzZWU0N2I2YmYwZmEwOGU5OTI4MWZmMCIsICJuZXh0X3VybCI6ICIifQ%3D%3D', version='HTTP/1.1', remote_ip='::ffff:127.0.0.1')
    Traceback (most recent call last):
      File "/Users/yuvipanda/.local/share/virtualenvs/oauthenticator/lib/python3.9/site-packages/tornado/web.py", line 1704, in _execute
        result = await result
      File "/Users/yuvipanda/code/oauthenticator/oauthenticator/oauth2.py", line 222, in get
        user = await self.login_user()
      File "/Users/yuvipanda/.local/share/virtualenvs/oauthenticator/lib/python3.9/site-packages/jupyterhub/handlers/base.py", line 804, in login_user
        authenticated = await self.authenticate(data)
      File "/Users/yuvipanda/.local/share/virtualenvs/oauthenticator/lib/python3.9/site-packages/jupyterhub/auth.py", line 473, in get_authenticated_user
        authenticated = await maybe_future(self.authenticate(handler, data))
      File "/Users/yuvipanda/code/oauthenticator/oauthenticator/github.py", line 249, in authenticate
        resp = await self.fetch(req, "fetching user teams", parse_json=False)
      File "/Users/yuvipanda/code/oauthenticator/oauthenticator/oauth2.py", line 381, in fetch
        raise e
      File "/Users/yuvipanda/code/oauthenticator/oauthenticator/oauth2.py", line 360, in fetch
        resp = await self.http_client.fetch(req, **kwargs)
    tornado.httpclient.HTTPClientError: HTTP 404: Not Found
```

This currently crashes the hub. I can probably handle this error (which we should do anyway) and provide a useful log message, but this reinforces to me that adding the scope if this feature is enabled is the right thing to do.

Reply:
I've removed the automatic scope setting in ac49a4c3fee761991464d71961b6bb03be25db03 and added docs.

Reply:
@manics reworked to use the Link header now!

Reply:
The question is probably between:
1. hard failure on startup if misconfigured.
2. warning in config that it has been added + docs it should be added
3. no warning and no suggestion in the docs to add read:org to the scopes

I lean away from the last, and initially thought i appreciated one the most but ended up thinking the second as best.

I prefer 2 over 1 since this cant be enforced in a clean way as for example using a schema for a helm chart - validated even before impacting the software operation. To have an admin disrupt a running hub with a hard error... Nah, not good

With 2, we have a purpose to allowing it - to avoid having an admin crash a deployment, but also with the docs, to help admins declare config that make sense and conveys how something works better.

Reply:
Let's continue this conversation elsewhere, I'll try open up something more holistic rather than piecemeal. I don't want you to have to spend the limited review energy you have on this, @consideRatio! It's more worthy reviewing the other parts :)

Reply:
Currently this PR adds docs + emits a warning, which is consistent with the behavior we have for it right now.